torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
             self.assertEqual(ddp_grads[0], local_model.fc.weight.grad)
             self.assertEqual(ddp_grads[1], local_model.fc.bias.grad)
 
+        def _test_hook_pickling(self, hook, hook_state):
+            learning_rate = 0.01
+            chkpt_file = tempfile.gettempdir() + "/checkpoint.pt"
+            rank = self.rank
+
+            input = torch.randn(1, 1, device=rank)
+            net = torch.nn.Linear(1, 5).to(rank)
+            ddp_model = torch.nn.parallel.DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            dummy_ddp_model = torch.nn.parallel.DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            optimizer = torch.optim.SGD(ddp_model.parameters(), lr=learning_rate)
+            ddp_model.register_comm_hook(hook_state, hook)
+
+            for _ in range(10):
+                optimizer.zero_grad()
+                loss = ddp_model(input).sum()
+                loss.backward()
+                optimizer.step()
+
+            state = {
+                'state_dict': ddp_model.state_dict(),
+                'comm_hook': pickle.dumps(hook),
+                'comm_hook_state': pickle.dumps(hook_state)
+            }
+
+            if rank == 0:
+                torch.save(state, chkpt_file)
+
+            dist.barrier()
+            map_location = {'cuda:%d' % 0: 'cuda:%d' % rank}
+            checkpoint = torch.load(chkpt_file, map_location=map_location)
+
+            dummy_ddp_model.load_state_dict(checkpoint['state_dict'])
+            dummy_hook = pickle.loads(checkpoint['comm_hook'])
+
+            with self.assertLogs() as captured:
+                dummy_hook_state = pickle.loads(checkpoint['comm_hook_state'])
+
+            # Check that the logger has only one entry
+            self.assertEqual(len(captured.records), 1)
+            # Check that the logger has an expected entry
+            self.assertEqual(
+                captured.records[0].getMessage(),
+                "NOTE: PowerSGDState is set to use the default process group."
+            )
+            # Check that loaded function is correct
+            self.assertEqual(dummy_hook.__qualname__, hook.__qualname__)
+
+            slots_original = hook_state.__slots__
+            slots_restored = dummy_hook_state.__slots__
+            self.assertEqual(slots_original, slots_restored)
+
+            slots = copy.copy(slots_restored)
```

Comment:
nit: Any reason to shallow copy the `list` and remove entries instead of using `if entry != "process_group" and entry != "rng"` when iterating over `slots_restored`?

Reply:
I agree that shallow copy is unnecessary. Updated according to suggestions.